File: python/Q300_longest_increasing_subsequence.py
#
# @lc app=leetcode id=300 lang=python3
#
# [300] Longest Increasing Subsequence
#

# @lc code=start
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        dp: list[int] = [1] * len(nums)
        result: int = 1

        for i in range(len(nums)):
            for j in range(i):
                if nums[j] < nums[i] and (new_value := dp[j] + 1) > dp[i]:
                    dp[i] = new_value

                    if new_value > result:
                        result = new_value

        return result


# @lc code=end

File: rust/src/leetcode/Q300_longest_increasing_subsequence.rs
/*
 * @lc app=leetcode id=300 lang=rust
 *
 * [300] Longest Increasing Subsequence
 */

// @lc code=start
impl Solution {
    pub fn length_of_lis(nums: Vec<i32>) -> i32 {
        let mut dp: Vec<i32> = vec![1; nums.len()];

        for i in 1..dp.len() {
            for j in 0..i {
                if nums[j] < nums[i] {
                    dp[i] = dp[i].max(dp[j] + 1);
                }
            }
        }

        *dp.iter().max().unwrap()
    }
}
// @lc code=end
